cmd/provisioner-localpv/app/helper_blockdevice.go

```diff
@@ -0,0 +1,214 @@
+
+This code was taken from https://github.com/rancher/local-path-provisioner
+and modified to work with the configuration options used by OpenEBS
+*/
+
+package app
+
+import (
+	//"fmt"
+	//"path/filepath"
+	//"strings"
+	"time"
+
+	"github.com/golang/glog"
+	//"github.com/pkg/errors"
+	errors "github.com/openebs/maya/pkg/errors/v1alpha1"
+
+	//hostpath "github.com/openebs/maya/pkg/hostpath/v1alpha1"
+
+	//container "github.com/openebs/maya/pkg/kubernetes/container/v1alpha1"
+	//pod "github.com/openebs/maya/pkg/kubernetes/pod/v1alpha1"
+	//volume "github.com/openebs/maya/pkg/kubernetes/volume/v1alpha1"
+	blockdevice "github.com/openebs/maya/pkg/blockdevice/v1alpha2"
+	blockdeviceclaim "github.com/openebs/maya/pkg/blockdeviceclaim/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	//ndmv1alpha1 "github.com/openebs/maya/pkg/apis/openebs.io/ndm/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	bdcStorageClassAnnotation = "local.openebs.io/blockdeviceclaim"
+)
+
+//TODO
+var (
+	//WaitForBDTimeoutCounts specifies the duration to wait for BDC to be associated with a BD
+	//The duration is the value specified here multiplied by 5
+	WaitForBDTimeoutCounts = 12
+)
+
+// HelperBlockDeviceOptions contains the options that
+// will launch a BDC on a specific node (nodeName)
+type HelperBlockDeviceOptions struct {
+	nodeName string
+	name     string
+	capacity string
+	//	deviceType string
+	bdcName string
+}
+
+// validate checks that the required fields to create BDC
+// are available
+func (blkDevOpts *HelperBlockDeviceOptions) validate() error {
+	glog.Infof("Validate Block Device Options")
+	if blkDevOpts.name == "" || blkDevOpts.nodeName == "" {
+		return errors.Errorf("invalid empty name or node")
+	}
+	return nil
+}
+
+// hasBDC checks if the bdcName has already been determined
+func (blkDevOpts *HelperBlockDeviceOptions) hasBDC() bool {
+	glog.Infof("Already has BDC %t", blkDevOpts.bdcName != "")
+	return blkDevOpts.bdcName != ""
+}
+
+// getBlcokDeviceClaimFromPV inspects the PV and fetches the BDC associated
+//  with the Local PV.
+func (blkDevOpts *HelperBlockDeviceOptions) setBlockDeviceClaimFromPV(pv *corev1.PersistentVolume) error {
+	glog.Infof("Setting Block Device Claim From PV")
+	bdc, found := pv.Annotations[bdcStorageClassAnnotation]
+	if found {
+		blkDevOpts.bdcName = bdc
+	}
+	return nil
+}
+
+// createBlockDeviceClaim creates a new BlockDeviceClaim for a given
+//  Local PV
+func (p *Provisioner) createBlockDeviceClaim(blkDevOpts *HelperBlockDeviceOptions) error {
+	glog.Infof("Creating Block Device Claim")
+	if err := blkDevOpts.validate(); err != nil {
+		return err
+	}
+
+	if blkDevOpts.hasBDC() {
+		//already created
+		glog.Infof("Volume %v has been initialized with BDC:%v", blkDevOpts.name, blkDevOpts.bdcName)
+		return nil
+	}
+
+	//Setup the BDC Name using the provided PV name.
+	//To help easily co-relate, the name format for bdc will
+	//be  "bdc-<pvname>"
+	bdcName := "bdc-" + blkDevOpts.name
```

Comment:
I am lost on how blockdevice's name is name of PV

Reply:
provided some more comments.